File: desktop/ClinicManagement.DesktopShell/ServerProbe.cs
using System;
using System.Net.Http;
using System.Net.Sockets;
using System.Security.Authentication;
using System.Threading;
using System.Threading.Tasks;

namespace ClinicManagement.DesktopShell;

/// <summary>
/// Settles which port a typed address means, when the user did not say.
///
/// <para>
/// The rule is identical in all three clients (desktop, Android, iOS) — see <c>mobile/CLAUDE.md</c> § « the port
/// rule ». An address with an explicit port is used verbatim and never probed. An address without one is tried
/// against <see cref="ServerConfig.CandidatePorts"/> in order, and the first port that **answers at all** wins.
/// </para>
///
/// <para>
/// ⚠️ « Answers » deliberately includes a TLS failure. An offline-LAN server presents a certificate signed by a CA
/// this PC may not have imported yet, so a handshake rejection is the *expected* outcome of probing a live clinic
/// server — treating it as "nothing here" would send every LAN install to the wrong port. What disqualifies a port
/// is a transport failure: no route, refused connection, timeout, or a name that does not resolve.
/// </para>
///
/// <para>
/// The mechanism is per-platform (here <see cref="HttpClient"/>, an <c>HttpURLConnection</c> on Android, a
/// <c>URLSession</c> on iOS) but the rule is not: what differs is how each runtime reports "the port answered",
/// never which port an address resolves to.
/// </para>
/// </summary>
public static class ServerProbe
{
    /// <summary>
    /// The route asked for. Anonymous and exempt from the client-version floor, so it answers a shell of any age —
    /// which is what makes it usable as a reachability probe rather than only as a version read.
    /// </summary>
    private const string ProbePath = "/api/meta/client-requirements";

    /// <summary>
    /// Per-candidate budget for <b>establishing the connection</b>. Short on purpose: this runs before the first
    /// paint of the connecting screen's navigation, and the worst case is paid once per address, not once per
    /// launch.
    /// </summary>
    private static readonly TimeSpan ConnectTimeout = TimeSpan.FromSeconds(3);

    /// <summary>
    /// Overall budget, deliberately <b>longer than <see cref="ConnectTimeout"/></b> — which is what makes the two
    /// phases distinguishable by exception type. Exceeding the connect budget throws
    /// <see cref="HttpRequestException"/>; exceeding this one throws <see cref="TaskCanceledException"/>. So a
    /// <c>TaskCanceledException</c> can only mean « the connection was established and the server then went quiet »,
    /// which is a live port. Were the two equal, the race between them would make that inference unsound.
    /// </summary>
    private static readonly TimeSpan OverallTimeout = TimeSpan.FromSeconds(5);

    /// <summary>
    /// The config to actually connect with. Returns <paramref name="config"/> unchanged when its port is already
    /// explicit, so the common case costs no network at all.
    /// </summary>
    /// <remarks>
    /// When **no** candidate answers, the first candidate is returned rather than nothing: the address is simply
    /// wrong or the server is off, and that is diagnosed far better by the navigation that follows — which shows
    /// the unreachable screen naming the address — than by a second error screen of this probe's own.
    /// </remarks>
    public static async Task<ServerConfig> ResolveAsync(ServerConfig config)
    {
        if (config.PortIsExplicit || !config.IsConfigured)
        {
            return config;
        }

        foreach (var port in config.CandidatePorts)
        {
            if (await AnswersAsync(config.Host, port).ConfigureAwait(true))
            {
                return config.WithResolvedPort(port);
            }
        }

        return config.WithResolvedPort(config.CandidatePorts[0]);
    }

    private static async Task<bool> AnswersAsync(string host, int port)
    {
        using var handler = new SocketsHttpHandler { ConnectTimeout = ConnectTimeout };
        using var client = new HttpClient(handler) { Timeout = OverallTimeout };
        try
        {
            // Any status is an answer — 200, 404 on a server too old to have the route, even a 502 from a proxy
            // in front of a starting API. All of them prove something is listening on this port.
            using var response = await client
                .GetAsync($"https://{host}:{port}{ProbePath}", HttpCompletionOption.ResponseHeadersRead)
                .ConfigureAwait(false);
            return true;
        }
        catch (HttpRequestException ex) when (IsTlsFailure(ex))
        {
            // A certificate this PC does not trust — the offline-LAN install's normal state before its CA is
            // imported. Something is listening and speaking TLS, which is all this probe asks.
            return true;
        }
        catch (TaskCanceledException)
        {
            // Connected, then silence until OverallTimeout — a listening but slow server. A managed host that
            // suspends an idle service accepts the connection at its edge immediately and only *then* wakes the
            // application, so the first response can take ten seconds or more (measured at 13.4 s against the live
            // Render front end). Reading that as « nothing on 443 » disqualifies the only port a hosted install
            // has. A connect that never completes cannot reach here: it throws HttpRequestException at
            // ConnectTimeout, two seconds earlier.
            return true;
        }
        catch (Exception)
        {
            // SocketException (refused / unreachable / DNS), a connect that exceeded ConnectTimeout, and anything
            // else unexpected: not a port to connect to.
            return false;
        }
    }

    private static bool IsTlsFailure(Exception exception)
    {
        for (var current = exception; current is not null; current = current.InnerException)
        {
            if (current is AuthenticationException)
            {
                return true;
            }

            if (current is SocketException)
            {
                return false;
            }
        }

        return false;
    }
}

File: mobile/android/app/src/main/java/com/clinicmanagement/shell/ServerProbe.kt
package com.clinicmanagement.shell

import android.util.Log
import java.io.IOException
import java.net.HttpURLConnection
import java.net.SocketTimeoutException
import java.net.URL
import java.net.UnknownHostException
import javax.net.ssl.SSLException

/**
 * Settles which port a typed address means, when the user did not say.
 *
 * The rule is identical in all three clients (desktop, Android, iOS) — see `mobile/CLAUDE.md` § « the port rule ».
 * An address with an explicit port is used verbatim and never probed. An address without one is tried against
 * [ServerConfig.candidatePorts] in order, and the first port that **answers at all** wins.
 *
 * ⚠️ « Answers » deliberately includes a TLS failure. An offline-LAN server presents a certificate signed by a CA
 * the phone may not have imported yet, so a handshake rejection is the *expected* outcome of probing a live clinic
 * server — treating it as « nothing here » would send every LAN install to the wrong port. What disqualifies a
 * port is a transport failure: no route, refused connection, a name that does not resolve, or a timeout **while
 * still connecting**.
 *
 * ⚠️ **A timeout while waiting for the answer is the second thing that counts as answering**, and it is the hosted
 * deployment's normal state rather than an edge case. A managed host that suspends an idle service accepts the TCP
 * connection at its edge immediately and only *then* wakes the application, so the connect succeeds in
 * milliseconds while the first response takes ten seconds or more — measured at **13.4 s** against the live Render
 * front end. Reading that as « nothing on 443 » disqualified the only port a hosted install has, fell through to
 * the LAN candidate, found nothing there either, and left the resolution to the not-found fallback below. The
 * question this probe asks is « is something listening on this port? », and a completed connection has already
 * answered it — so the phase the timeout happened in is what decides, not the timeout itself.
 *
 * ⚠️ Blocking I/O, exactly like [ClientRequirements.fetch], and called from the same background executor. It runs
 * *before* that fetch because the fetch needs a base URL, and until this has run there is no base URL to give it.
 */
object ServerProbe {

    private const val TAG = "ClinicShell"

    /**
     * The route asked for. Anonymous and exempt from the client-version floor, so it answers a shell of any age —
     * which is what makes it usable as a reachability probe rather than only as a version read.
     */
    private const val PATH = "/api/meta/client-requirements"

    /**
     * Per-candidate budget. Short on purpose: this runs behind the « Connexion… » screen, and the worst case is
     * paid once per address, not once per launch.
     *
     * Both stay short, including the read: a slow answer no longer needs to be *waited* for, because the connect
     * having succeeded is already the whole verdict. Raising the read budget to outlast a cold start would make
     * every genuinely dead candidate cost that same budget, which is the trade the 443-before-5001 ordering exists
     * to avoid.
     */
    private const val CONNECT_TIMEOUT_MS = 4_000
    private const val READ_TIMEOUT_MS = 4_000

    /**
     * The config to actually connect with. Returns [config] unchanged when its port is already explicit, so the
     * common case costs no network at all.
     *
     * When **no** candidate answers, the first candidate is returned rather than nothing: the address is simply
     * wrong or the server is off, and that is diagnosed far better by the load that follows — which shows the
     * unreachable screen naming the address — than by a second error state of this probe's own.
     */
    fun resolve(config: ServerConfig): ServerConfig {
        if (config.portIsExplicit || !config.isConfigured) return config

        for (port in config.candidatePorts) {
            if (answers(config.host, port)) {
                Log.i(TAG, "Server address resolved to port $port")
                return config.withResolvedPort(port)
            }
        }

        return config.withResolvedPort(config.candidatePorts.first())
    }

    private fun answers(host: String, port: Int): Boolean {
        var connection: HttpURLConnection? = null
        // Which phase a timeout lands in is the whole verdict, and `SocketTimeoutException` does not say. So the
        // connect is made explicitly and the fact recorded: anything that times out after this is true was waiting
        // on a server that had already accepted it.
        var connected = false
        return try {
            connection = (URL("https://$host:$port$PATH").openConnection() as HttpURLConnection).apply {
                requestMethod = "GET"
                connectTimeout = CONNECT_TIMEOUT_MS
                readTimeout = READ_TIMEOUT_MS
                setRequestProperty("Accept", "application/json")
            }
            connection.connect()
            connected = true
            // Any status is an answer — 200, 404 on a server too old to have the route, even a 502 from a proxy
            // in front of a starting API. All of them prove something is listening on this port.
            connection.responseCode
            true
        } catch (e: SSLException) {
            // A certificate the phone does not trust — the offline-LAN install's normal state before its CA is
            // imported. Something is listening and speaking TLS, which is all this probe asks. Thrown out of
            // `connect()`, so `connected` is still false and this branch is what has to answer.
            true
        } catch (e: UnknownHostException) {
            false
        } catch (e: SocketTimeoutException) {
            // Connected then silent ⇒ listening but slow (a suspended hosted service waking up) ⇒ answers.
            // Timed out before connecting ⇒ nothing reachable on this port ⇒ does not answer.
            connected
        } catch (e: IOException) {
            false
        } finally {
            connection?.disconnect()
        }
    }
}
